crates/sui-graphql-client/src/lib.rs

```diff
@@ -0,0 +1,584 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#![doc = include_str!("../README.md")]
+
+pub mod query_types;
+
+use base64ct::Encoding;
+use query_types::{
+    ChainIdentifierQuery, CheckpointArgs, CheckpointId, CheckpointQuery, CoinMetadata,
+    CoinMetadataArgs, CoinMetadataQuery, EpochSummaryArgs, EpochSummaryQuery, EventFilter,
+    EventsQuery, EventsQueryArgs, ObjectFilter, ObjectQuery, ObjectQueryArgs, ObjectsQuery,
+    ObjectsQueryArgs, PageInfo, ProtocolConfigQuery, ProtocolConfigs, ProtocolVersionArgs,
+    ServiceConfig, ServiceConfigQuery, TransactionBlockArgs, TransactionBlockQuery,
+    TransactionBlocksQuery, TransactionBlocksQueryArgs, TransactionsFilter, Uint53,
+};
+use sui_types::types::{
+    Address, CheckpointSequenceNumber, CheckpointSummary, Event, Object, SignedTransaction,
+};
+
+use anyhow::{anyhow, ensure, Error, Result};
+use cynic::{serde, GraphQlResponse, Operation, QueryBuilder};
+
+const MAINNET_HOST: &str = "https://sui-mainnet.mystenlabs.com/graphql";
+const TESTNET_HOST: &str = "https://sui-testnet.mystenlabs.com/graphql";
+const DEVNET_HOST: &str = "https://sui-devnet.mystenlabs.com/graphql";
+const DEFAULT_LOCAL_HOST: &str = "http://localhost:9125/graphql";
+
+#[derive(Debug)]
+/// A page of items returned by the GraphQL server.
+pub struct Page<T> {
+    /// Information about the page, such as the cursor and whether there are more pages.
+    page_info: PageInfo,
+    /// The data returned by the server.
+    data: Vec<T>,
+}
+
+impl<T> Page<T> {
+    pub fn page_info(&self) -> &PageInfo {
+        &self.page_info
+    }
+
+    pub fn data(&self) -> &[T] {
+        &self.data
+    }
+
+    fn new(page_info: PageInfo, data: Vec<T>) -> Self {
+        Self { page_info, data }
+    }
+}
+
+/// The Client is a GraphQL client for interacting with the Sui blockchain.
+/// By default, it uses Reqwest as the HTTP client.
+pub struct Client {
+    /// The URL of the GraphQL server.
+    rpc: String,
+    /// GraphQL supports multiple versions of the service. By default, the client uses the stable
+    /// version of the service.
+    rpc_version: Option<String>,
+    inner: reqwest::Client,
+}
+
+impl Client {
+    // ===========================================================================
+    // Client Misc API
+    // ===========================================================================
+
+    /// Create a new GraphQL client with the provided server address.
+    pub fn new(server: &str) -> Result<Self, Error> {
+        reqwest::Url::parse(server).map_err(|_| anyhow!("Invalid URL: {}", server))?;
+
+        let client = Client {
+            rpc: server.to_string(),
+            rpc_version: None,
+            inner: reqwest::Client::new(),
```

Comment:
could we add a default UserAgent header to the client? will make it easier to track in the future.

Reply:
Added.